### ISSUE TITLE
Consuming python context managers

### ISSUE THREAD

**Issue by olivercoad** [fable-compiler/Fable.Python#25]
Trying to work out how to do the equivalent of

```python
with open("README.md", "r") as fh:
    long_description = fh.read()
```

First, is `read` in builtins meant to be `open`?

https://github.com/fable-compiler/Fable.Python/blob/4634758a36a34ebfad1ae7fb1cea1b48321eb895/src/stdlib/Builtins.fs#L118-L128

Second, what's the best way of ensuring fh is closed?
It would be nice if f# `use` could be compiled to python `with`.
Not sure if that would be possible though without a complete solution for IDisposable that also handles translating a direct call to Dispose.
https://docs.python.org/3/reference/compound_stmts.html#the-with-statement

My current solution:

```fsharp
open Fable.Core

type OpenFile =
    abstract member read : unit -> string
    abstract member close : unit -> unit

[<Emit("open($0,$1)")>]
let open' (filename:string, mode:string) : OpenFile = nativeOnly

do
    let long_description =
        let fh = open' ("README.md", "r")
        try
            fh.read()
        finally
            fh.close()
```

**Comment by dbrattli** [comment]
Thanks for submitting. I'm looking into this

**Comment by dbrattli** [comment]
Fixing by renaming to ` ``open`` `, which translates directly without needing to `Emit`. 

For disposing, then the next Fable compiler will allow disposing of the Resource Manager protocol i.e `__exit__` if `Dispose` is not available. It would be nice to compile to Python `with`  but it turns out that it's quite difficult since F# `use` is not included in the F# AST. It's represented as try-finally, thus the value assignment is somewhere before that in the AST which makes it difficult to find it with a pattern match.

**Comment by olivercoad** [comment]
Marking interfaces with IDisposable and calling `__exit__` if `Dispose` isn't available sounds like a nice solution, but I'm concerned that it might only work for "simple" context managers do nothing in `__enter__` and just `return self`.
Maybe it'll work fine for `open`, but when looking at https://www.python.org/dev/peps/pep-0343/ there are many examples where `__enter__` can't be ignored.

**Comment by dbrattli** [comment]
@olivercoad Yes, you are right. We need to make it better. Trying to find the right place to pattern match this. It might be possible to for `Let` when the type is `IDisposable` and the first item in the body is a `TryFinally`, and the finally is disposing, e.g: (choose typed ast) https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWQ2wAuYVYbAEkAIkQAOrNFQBG3MAF4wAV2QzWADzDJlNGgB1kh7nx7o%2BACgCUiwxAjL6YGYv537MgHRi0k%2BtcgQAF8gA

**Comment by olivercoad** [comment]
I don't think it's useful to match `use` and emit `with` until we have a solution broader issue of the mismatch between IDisposable and context managers. Using `with` rather than the more verbose equivalent would be a nice-to-have, but IMO getting the semantics right first is more important.

As I see it, there seems to be a fundamental difference between IDisposable in .NET and context managers in python., and that is that in .NET, once you have an instance of an `IDisposable`, it's ready to use and but in python you

**Comment by olivercoad** [comment]
I don't think it's useful to match `use` and emit `with` until we have a solution broader issue of the mismatch between IDisposable and context managers. Using `with` rather than the more verbose equivalent would be a nice-to-have, but IMO getting the semantics right first is more important.

As I see it, there seems to be a fundamental difference between IDisposable in .NET and context managers in python. In .NET, once you have an instance of an `IDisposable`, it's more or less ready to use, and you just have to dispose it when done. But in python, the context manager handles initialising as well as disposing, and is (sometimes) a different object to the one the caller is interested in.

Basically what I'm saying is that if we just blindly compile `use` to `with`, it would work for "simple" context managers, but other cases could have nasty gotchas:
```
use a = some_expression()
let b = some_expression()
type(a) <> type(b)
```

This is just my reasoning based on looking at the python with-statement docs; I'm not experienced with python at all, so could be completely wrong.

**Comment by dbrattli** [comment]
I think the trick here is to make every `IDisposable` a simple context manager with an `__enter__` that returns `self`.  When generating a class that implements `IDisposable` we also need to make sure we add the context manager:

```py
class IDisposable:
    @abstractmethod
    def Dispose(self) -> None:
        ...

    def __enter__(self):
        """Enter context management."""
        return self

    def __exit__(
        self, exctype: Optional[Type[BaseException]], excinst: Optional[BaseException], exctb: Optional[TracebackType]
    ) -> bool:
        """Exit context management."""

        self.Dispose()
        return False
```

So for a compiled `IDisposable` then `a` and `b` would be the same. For a Python context manager it may not. Thus all existing F# code would behave the same as before, but Python bindings may give you a different `a`  when you use `use`.

**Comment by dbrattli** [comment]
But we still have the problem of the returned object being possibly two different types since not using `with` in Python will not trigger enter/exit i.e `_mgr` vs `_mgr.__enter__()`

```fs
// Should not use context manager
let fh = open' ("README.md", "r")
fh.Dispose()
```
vs
```fs
// Should use context manager
use fh = open' ("README.md", "r")
///
```

So we cannot make both type check in F# using an attribute and we should perhaps not force to use the context manager (or not). 

Another possibility is using an explicit `ContextManager` property that erases in Python, so we can get (possibly) a different type in F#.

```fs
type OpenFileCM =
    inherit IDisposable

    abstract member read : unit -> string
    abstract member close : unit -> unit

type OpenFile =
    inherit IDisposable

    abstract member read : unit -> string
    abstract member close : unit -> unit

    [<Emit("$0")>]
    abstract member ContextManager: OpenFileCM

[<Emit("open($0,$1)")>]
let open' (filename:string, mode:string) : OpenFile = nativeOnly

let test () =
    use fh = open'("README.md", "r").ContextManager
    fh.read()
    // ...
    fh.Dispose() // Early dispose (if any)
```

This should then generate the following Python code:

```py
from fable_modules.fable_library.util import dispose

def test():
    _mgr = open("README.md","r")
    with _mgr as fh:
        fh.read()
        dispose(_mgr) #  Early dispose (if any). Calls .Dispose() or .__exit__()
```

or just generate separate function/methods to be used with use bindings, to easier spot it with dot-completions:

```fs
/// open to be used with `let` bindings.
let ``open``(filename:string, mode:string) : OpenFile = nativeOnly
/// open to be used with `use` bindings.
let open_cm (filename:string, mode:string) : OpenFileCM = nativeOnly
```

**Comment by olivercoad** [comment]
```fsharp
// Should not use context manager
let fh = open' ("README.md", "r")
fh.Dispose()
```

Why not? If it's not using the context manager, then what should Dispose do?

What I'm suggesting is that methods anotated with `ContextManager` will always be forced to use it even with `let`, so the calling code never gets a handle on the context manager (so it doesn't need a type), only the returned object from `__enter__`.

**Comment by dbrattli** [comment]
Because you might want to hold the context manager without activating the protocol, e.g for passing it around etc. The example below will never call enter/exit. I think it would be wrong if we did.
```py
class OpenFile:
    def __enter__(self) -> "OpenFile":
        print("enter")
        return self

    def __exit__(self, a, b, c) -> None:
        print("exit")

    def __repr__(self) -> str:
        return ""

    def close(self) -> None:
        pass


m = OpenFile()
m.close()
```

**Comment by olivercoad** [comment]
> It makes the fh hot (vs cold)

That was kinda my intention, because in .NET usually `IDisposable` *is* "hot".

> you might want to hold the context manager without activating the protocol

In that case when writing F# bindings you have to aware of the way the context manager works and declare two types if necessary, like you say.

> using an explicit ContextManager property that erases in Python

Can you show how that would work for a less trivial context manager?

```python
class contrived_example:
    def __init__(self, filename):
        self.filename = filename
    def __enter__(self):
        self.f = open(self.filename, "r")
        return enumerate(self.f)
    def __exit__(self, type, value, tb):
        self.f.close()

# used as follows
with contrived_example("/etc/passwd") as enumerator:
    for (lnum, line) in enumerator:
        print(str(lnum) + ": " + line.rstrip())
```

**Comment by olivercoad** [comment]
This example highlights a problem with `ContextManagerAttribute` because the return type of `__enter__` doesn't inherit `IDisposable`.

**Comment by dbrattli** [comment]
(I'm btw gotten sick, I'll be back in a few days)

**Comment by dbrattli** [comment]
This should (mostly) be fixed so closing. If it still fails for you then please open a new issue with a failing example/test.
